SignalServiceKit/src/Messages/OWSMessageSender.m

```diff
@@ -1151,38 +1151,34 @@ - (void)handleSendToMyself:(TSOutgoingMessage *)outgoingMessage
+        TSContactThread *cThread = [TSContactThread getOrCreateThreadWithContactId:contactId transaction:transaction];
         [cThread saveWithTransaction:transaction];
 
         // We need to clone any attachments for message sent to self; otherwise deleting
         // the incoming or outgoing copy of the message will break the other.
-        NSMutableArray<NSString *> *attachmentIds = [NSMutableArray new];
+        NSMutableArray<NSString *> *incomingAttachmentIds = [NSMutableArray new];
         for (NSString *attachmentId in outgoingMessage.attachmentIds) {
-            TSAttachmentStream *_Nullable outgoingAttachment =
-                [TSAttachmentStream fetchObjectWithUniqueID:attachmentId transaction:transaction];
-            OWSAssert(outgoingAttachment);
-            if (!outgoingAttachment) {
-                DDLogError(@"%@ Couldn't load outgoing attachment for message sent to self.", self.logTag);
+            TSAttachment *_Nullable incomingAttachment =
+                [self cloneAttachmentForSendToMyself:attachmentId transaction:transaction];
+            if (incomingAttachment) {
+                [incomingAttachmentIds addObject:incomingAttachment.uniqueId];
             } else {
-                TSAttachmentStream *incomingAttachment =
-                    [[TSAttachmentStream alloc] initWithContentType:outgoingAttachment.contentType
-                                                          byteCount:outgoingAttachment.byteCount
-                                                     sourceFilename:outgoingAttachment.sourceFilename];
-                NSError *error;
-                NSData *_Nullable data = [outgoingAttachment readDataFromFileWithError:&error];
-                if (!data || error) {
-                    DDLogError(@"%@ Couldn't load attachment data for message sent to self: %@.", self.logTag, error);
-                } else {
-                    [incomingAttachment writeData:data error:&error];
-                    if (error) {
-                        DDLogError(
-                            @"%@ Couldn't copy attachment data for message sent to self: %@.", self.logTag, error);
-                    } else {
-                        [incomingAttachment saveWithTransaction:transaction];
-                        [attachmentIds addObject:incomingAttachment.uniqueId];
-                    }
-                }
+                DDLogError(@"%@ couldn't clone outgoing attachment for send to self.", self.logTag);
+            }
+        }
+
+        // TODO: Copy attachments for quoted messages.
+        TSQuotedMessage *_Nullable incomingQuotedMessage = [outgoingMessage.quotedMessage copy];
+
+        OWSContact *_Nullable incomingContactShare = [outgoingMessage.contactShare copy];
+        if (incomingContactShare.avatarAttachmentId) {
+            TSAttachment *_Nullable incomingAttachment =
+                [self cloneAttachmentForSendToMyself:incomingContactShare.avatarAttachmentId transaction:transaction];
```

Comment:
We also use `cloneAttachmentForSendToMyself` to clone contact share attachments.